jlm/util/GraphWriter.hpp

```diff
@@ -0,0 +1,843 @@
+   * @param object a pointer to a program object, must be non-null.
+   */
+  void
+  SetAttributeObject(const std::string & attribute, void * object);
+
+  /**
+   * Assigns or overrides a given attribute on the element with a reference to a graph element.
+   * This allows associations between graph elements to be included in the output, across graphs.
+   * The element must be a part of the same GraphWriter instance.
+   * @param attribute the name of the attribute.
+   * @param element the graph element whose id should be used as attribute value.
+   */
+  void
+  SetAttributeGraphElement(const std::string & attribute, GraphElement & element);
+
+  /**
+   * Claims a unique id suffix for the element, if it doesn't already have one.
+   */
+  virtual void
+  Finalize();
+
+  /**
+   * @return true if this GraphElement has been finalized, otherwise false.
+   */
+  [[nodiscard]] bool
+  IsFinalized() const;
+
+  /**
+   * Prints the attributes of the graph element.
+   * @param out the stream output is written to.
+   * @param format the output format to use.
+   */
+  void
+  OutputAttributes(std::ostream & out, AttributeOutputFormat format);
+
+private:
+  void
+  SetProgramObjectUintptr(uintptr_t object);
+
+  // A human-readable piece of text that should be rendered with the element
+  std::string Label_;
+
+  // A number added to the end of the id stub to make it globally unique
+  int UniqueIdSuffix_;
+
+  // The object in the program this graph object corresponds to, or 0 if none
+  uintptr_t ProgramObject_;
+
+  // Arbitrary collection of other attributes. The value can be a string, a reference to a
+  // GraphElement, or a reference to a program object.
+  using AttributeValue = std::variant<std::string, GraphElement *, uintptr_t>;
+  std::unordered_map<std::string, AttributeValue> AttributeMap_;
+};
+
+/**
+ * Abstract class representing a part of a node that edges can be attached to
+ */
+class Port : public GraphElement
+{
+  friend Edge;
+
+protected:
+  Port();
+
+public:
+  ~Port() override = default;
+
+  virtual Node &
+  GetNode() = 0;
+
+  [[nodiscard]] Graph &
+  GetGraph() override;
+
+  /**
+   * @return true if a directed edge may have its head at this port, false otherwise
+   */
+  [[nodiscard]] virtual bool
+  CanBeEdgeHead() const;
+
+  /**
+   * @return true if a directed edge may have its tail at this port, false otherwise
+   */
+  [[nodiscard]] virtual bool
+  CanBeEdgeTail() const;
+
+  /**
+   * @return true if any edges are leaving this port, or any non-directed edges are present
+   */
+  [[nodiscard]] bool
+  HasOutgoingEdges() const;
+
+  /**
+   * @return true if any edges are pointing at this port, or any non-directed edges are present
+   */
+  [[nodiscard]] bool
+  HasIncomingEdges() const;
+
+  /**
+   * Helper function for setting the background color of the Port using the correct dot attributes
+   */
+  virtual void
+  SetFillColor(std::string color) = 0;
```

Comment:
Could we use an enum instead of a string here?

Reply:
The color can be any hex value, so it might make more sense to provide a list of common colors, but not force the parameter to be an enum?

Reply:
Sounds good to me.